src/Libraries/Microsoft.Extensions.Telemetry/Buffering/GlobalBuffer.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#if NET9_0_OR_GREATER
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Diagnostics;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Extensions.Logging;
+
+internal sealed class GlobalBuffer : BackgroundService, ILoggingBuffer
+{
+    private readonly IOptionsMonitor<GlobalBufferOptions> _options;
+    private readonly ConcurrentDictionary<IBufferedLogger, ConcurrentQueue<GlobalBufferedLogRecord>> _buffers;
+    private readonly TimeProvider _timeProvider = TimeProvider.System;
+    private DateTimeOffset _lastFlushTimestamp;
+
+    public GlobalBuffer(IOptionsMonitor<GlobalBufferOptions> options)
+    {
+        _options = options;
+        _lastFlushTimestamp = _timeProvider.GetUtcNow();
+        _buffers = new ConcurrentDictionary<IBufferedLogger, ConcurrentQueue<GlobalBufferedLogRecord>>();
+    }
+
+    internal GlobalBuffer(IOptionsMonitor<GlobalBufferOptions> options, TimeProvider timeProvider)
+        : this(options)
+    {
+        _timeProvider = timeProvider;
+        _lastFlushTimestamp = _timeProvider.GetUtcNow();
+    }
+
+    public bool TryEnqueue(
+        IBufferedLogger logger,
+        LogLevel logLevel,
+        string category,
+        EventId eventId,
+        IReadOnlyList<KeyValuePair<string, object?>> joiner,
+        Exception? exception, string formatter)
+    {
+        if (!IsEnabled(category, logLevel, eventId))
+        {
+            return false;
+        }
+
+        var record = new GlobalBufferedLogRecord(logLevel, eventId, joiner, exception, formatter);
+        var queue = _buffers.GetOrAdd(logger, _ => new ConcurrentQueue<GlobalBufferedLogRecord>());
+        if (queue.Count >= _options.CurrentValue.Capacity)
+        {
+            _ = queue.TryDequeue(out GlobalBufferedLogRecord? _);
+        }
+
+        queue.Enqueue(record);
+
+        return true;
+    }
+
+    public void Flush()
+    {
+        foreach (var (logger, queue) in _buffers)
```

Comment:
You and Martin should decide if you care about emitting the events in any particular order. As-is it looks like the sort order is first by an arbitrary category ordering and secondly by approximate time-order. If users see buffered logs dumped on the console that way they might be confused by it but I imagine most logging backends won't care at all. I'm fine with it as-is but I wanted to raise the question to ensure everyone else also is comfortable with it.